apm-sniffer/apm-sdk-plugin/pulsar-plugin/src/main/java/org/apache/skywalking/apm/plugin/pulsar/PulsarProducerInterceptor.java

```diff
@@ -0,0 +1,85 @@
+import org.apache.skywalking.apm.agent.core.context.ContextSnapshot;
+import org.apache.skywalking.apm.agent.core.context.tag.Tags;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+
+import java.lang.reflect.Method;
+
+public class PulsarProducerInterceptor implements InstanceMethodsAroundInterceptor {
+
+    public static final String OPERATE_NAME_PREFIX = "Pulsar/";
+    public static final String PRODUCER_OPERATE_NAME_SUFFIX = "/Producer";
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                             MethodInterceptResult result) throws Throwable {
+        ProducerEnhanceRequiredInfo requiredInfo = (ProducerEnhanceRequiredInfo) objInst.getSkyWalkingDynamicField();
+        ContextCarrier contextCarrier = new ContextCarrier();
+        String topicName = requiredInfo.getTopic();
+        AbstractSpan activeSpan = ContextManager.createExitSpan(OPERATE_NAME_PREFIX + topicName +
+                PRODUCER_OPERATE_NAME_SUFFIX, contextCarrier, requiredInfo.getServiceUrl());
+        Tags.MQ_BROKER.set(activeSpan, requiredInfo.getServiceUrl());
+        Tags.MQ_TOPIC.set(activeSpan, topicName);
+        SpanLayer.asMQ(activeSpan);
+        activeSpan.setComponent(ComponentsDefine.PULSAR_PRODUCER);
+        CarrierItem next = contextCarrier.items();
+        MessageImpl msg = (MessageImpl) allArguments[0];
+        while (next.hasNext()) {
+            next = next.next();
+            msg.getMessageBuilder().addProperties(PulsarApi.KeyValue.newBuilder()
+                    .setKey(next.getHeadKey())
+                    .setValue(next.getHeadValue()));
+        }
+        EnhancedInstance callbackInstance = (EnhancedInstance) allArguments[1];
+        if (callbackInstance != null) {
+            ContextSnapshot snapshot = ContextManager.capture();
+            if (null != snapshot) {
+                SendCallbackEnhanceRequiredInfo callbackRequiredInfo = new SendCallbackEnhanceRequiredInfo();
+                callbackRequiredInfo.setTopic(topicName);
+                callbackRequiredInfo.setContextSnapshot(snapshot);
+                callbackInstance.setSkyWalkingDynamicField(callbackRequiredInfo);
+            }
+        }
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                              Object ret) throws Throwable {
+        ContextManager.stopSpan();
```

Comment:
I think we need to ensure SPAN has been created will be better.

Reply:
As mentioned above, to catch the IllegalStateException or is their any method to check the span is created

Reply:
I think if logically you could generate the span created in Pulsar plugin, it is OK to just stop span directly.

Because even we do `isActive` check, you could just check the span created by other plugins.